src/System.Private.CoreLib/src/System/Threading/ThreadPool.Unix.cs

```diff
@@ -4,18 +4,206 @@
+
+        private bool _signalAfterCallbacksComplete;
+
+        private int _unregisterCalled;
+
+        private AutoResetEvent _unregisteredEvent = new AutoResetEvent(false);
+
+        internal bool Unregister(WaitHandle waitObject)
+        {
+            if (Interlocked.Exchange(ref _unregisterCalled, 1) == 0)
+            {
+                UserUnregisterWaitHandle = waitObject?.SafeWaitHandle;
+                if (UserUnregisterWaitHandle != null && !UserUnregisterWaitHandle.IsInvalid)
+                {
+                    UserUnregisterWaitHandle.DangerousAddRef();
+                    UserUnregisterWaitHandleValue = UserUnregisterWaitHandle.DangerousGetHandle();
+                    IsBlocking = UserUnregisterWaitHandleValue == new IntPtr(-1);
+                }
+                else
+                {
+                    IsBlocking = false;
+                }
+
+                if (_unregisterSignaled == 0)
+                {
+                    WaitThread.UnregisterWait(this);
+                }
+                return true;
+            }
+            return false;
+        }
+
+        /// <summary>
+        /// Signal <see cref="UserUnregisterWaitHandle"/> if it has not been signaled yet and is a valid handle.
+        /// </summary>
+        private void SignalUserWaitHandle()
+        {
+            if (Interlocked.Exchange(ref _unregisterSignaled, 1) == 0)
+            {
+                SafeWaitHandle handle = UserUnregisterWaitHandle;
+                if (!handle.IsInvalid)
+                {
+                    try
+                    {
+                        EventWaitHandle.Set(handle);
+                    }
+                    finally
+                    {
+                        handle.DangerousRelease();
+                    }
+                }
+                _unregisteredEvent.Set();
+            }
+        }
+
+        /// <summary>
+        /// Perform the registered callback if the <see cref="UserUnregisterWaitHandle"/> has not been signaled.
+        /// </summary>
+        /// <param name="timedOut">Whether or not the wait timed out.</param>
+        internal void PerformCallback(bool timedOut)
+        {
+            if (_unregisterSignaled == 0)
+            {
+                _ThreadPoolWaitOrTimerCallback.PerformWaitOrTimerCallback(Callback, timedOut);
+            }
+            CompleteCallbackRequest();
+        }
+
+        internal void RequestCallback()
+        {
+            _callbackLock.Acquire();
+            try
+            {
+                _numRequestedCallbacks++;
+            }
+            finally
+            {
+                _callbackLock.Release();
+            }
+        }
+
+        internal void TrySignalUserWaitHandle()
+        {
+            _callbackLock.Acquire();
+            try
+            {
+
+                if (UserUnregisterWaitHandle != null && UserUnregisterWaitHandleValue != new IntPtr(-1))
```

Comment:
I think this check should be removed, and other things need fixing as well. For instance, consider the following order of events:
- A non-repeating wait is registered
- QueueWaitCompletion registers the removal and queues a callback
- This function gets called and since Unregister hasn't been called yet, it doesn't do anything due to the above check
- A blocking Unregister is called, and it waits for the callbacks to complete in WaitThread.UnregisterWait
- The callback completes. CompleteCallbackRequest gets called but does not signal the appropriate event to release WaitThread.UnregisterWait because _signalAfterCallbacksComplete is false.

It could probably be fixed in several ways, but considering my other comments, I think the following works:

SignalUserWaitHandle:
  - Could be considered to have one of two jobs
    - For a non-blocking Unregister with a valid event, signal the user event
    - For a blocking Unregister with a -1 event handle value, signal the event that releases WaitThread.UnregisterWait
  - If you use this solution, consider renaming the function to OnAllCallbacksComplete or something like that since it doesn't just signal the user wait handle

In this function, if _numRequestedCallbacks is not 0, it should set _signalAfterCallbacksComplete to true regardless of the user wait handle so that callback completion can release WaitThread.UnregisterWait.

In a slight modification of the above example, if Unregister is called after the callback completes, then Unregister itself should be able to identify that case without any races and signal the user event directly if need be.